src/platform/packages/shared/kbn-esql-validation-autocomplete/src/validation/validation.ts

```diff
@@ -109,51 +105,68 @@ async function validateAst(
 
-  const [sources, availableFields, availablePolicies, joinIndices] = await Promise.all([
+  const [sources, availablePolicies, joinIndices] = await Promise.all([
     // retrieve the list of available sources
     retrieveSources(rootCommands, callbacks),
-    // retrieve available fields (if a source command has been defined)
-    retrieveFields(queryString, rootCommands, callbacks),
     // retrieve available policies (if an enrich command has been defined)
     retrievePolicies(rootCommands, callbacks),
     // retrieve indices for join command
     callbacks?.getJoinIndices?.(),
   ]);
 
-  if (availablePolicies.size) {
-    const fieldsFromPoliciesMap = await retrievePoliciesFields(
-      rootCommands,
-      availablePolicies,
-      callbacks
-    );
-    fieldsFromPoliciesMap.forEach((value, key) => availableFields.set(key, value));
-  }
+  const sourceQuery = queryString.split('|')[0];
+  const sourceFields = await getColumnsByTypeHelper(
+    EsqlQuery.fromSrc(sourceQuery).ast,
+    sourceQuery,
+    callbacks
+  ).getColumnMap();
 
-  const userDefinedColumns = collectUserDefinedColumns(rootCommands, availableFields, queryString);
-  messages.push(...validateUnsupportedTypeFields(availableFields, rootCommands));
-
-  const references: ReferenceMaps = {
-    sources,
-    fields: availableFields,
-    policies: availablePolicies,
-    userDefinedColumns,
-    query: queryString,
-    joinIndices: joinIndices?.indices || [],
-  };
+  messages.push(
+    ...validateUnsupportedTypeFields(
+      sourceFields as Map<string, ESQLFieldWithMetadata>,
+      rootCommands
+    )
+  );
 
   const license = await callbacks?.getLicense?.();
   const hasMinimumLicenseRequired = license?.hasAtLeast;
-  for (const [_, command] of rootCommands.entries()) {
-    const commandMessages = validateCommand(command, references, rootCommands, {
-      ...callbacks,
-      hasMinimumLicenseRequired,
-    });
+
+  /**
+   * Even though we are validating single commands, we work with subqueries.
+   *
+   * The reason is that building the list of columns available in each command requires
+   * the full command subsequence that precedes that command.
+   */
+  const subqueries = getSubqueriesToValidate(rootCommands);
+  for (const subquery of subqueries) {
+    const { getColumnMap } = getColumnsByTypeHelper(subquery, queryString, callbacks);
+    const availableColumns = await getColumnMap();
+
+    const references: ReferenceMaps = {
+      sources,
+      columns: availableColumns,
+      policies: availablePolicies,
+      query: queryString,
+      joinIndices: joinIndices?.indices || [],
+    };
+
+    const commandMessages = validateCommand(
+      subquery.commands[subquery.commands.length - 1],
+      references,
+      rootCommands,
+      {
+        ...callbacks,
+        hasMinimumLicenseRequired,
+      }
+    );
     messages.push(...commandMessages);
   }
 
   const parserErrors = parsingResult.errors;
 
   /**
+   * @TODO - move deeper
```

Comment:
Philosophical

Reply:
Gone in https://github.com/elastic/kibana/pull/233221/commits/a6857e6a7f6abb73a76c47b2b4bfec62794bd31a